### PR TITLE
Special-case fsdp wrapped modules to be Unspecialized

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89457
* #89802
* #89520
* #89523
* __->__ #89330
### Summary
Making dynamo treat the nn.Modules inside FSDP wrappers as 'Unspecialized'
results in dynamo-produced graphs where nn.module parameters are inputs
to the graph rather than attributes of the outer graphmodule.

This helps in FSDP since it forces dynamo to pick the latest copy
of the parameters off the user's nn.Module (which FSDP mutates every pre_forward),
solving the ordering issue in backward.

### Details
Imagine this toy model
```
class MyModule(torch.nn.Module):
    def __init__(self, a, b):
        super(MyModule, self).__init__()
        self.net = nn.Sequential(
            nn.Linear(a, b),
            nn.ReLU(),
        )
    def forward(self, x):
        return self.net(x)

class ToyModel(nn.Module):
    def __init__(self):
        super(ToyModel, self).__init__()
        self.net = nn.Sequential(
            *[MyModule(10, 10000)]
            + [MyModule(10000, 1000)]
            + [MyModule(1000, 5)]
        )

    def forward(self, x):
        return self.net(x)
```
Where FSDP is recursively wrapped around each `MyModule`, then dynamo-compiled, with dynamo already configured to skip/break in FSDP code.  You'd expect to get 3 compiled AOT functions, corresponding to the contents of `MyModule`, and then see FSDP's communication ops happen inbetween them (eagerly).  This almost happens (everything works out fine in forward), but in backward there is an ordering issue.

FSDP creates a flat buffer for all the parameters that are bucketed together, and then creates views into this buffer to replace the original parameters.  On each iteration of forward, it creates a new view after 'filling' the flatbuffer with data from an all-gather operation, to 'unshard' the parameters from remote devices.  Dynamo traces the first such view and stores it in a compiled graphmodule.

During  tracing, we see (1) view created for first MyModule, (2) compile first MyModule, (3) ... for the rest of layers

Then during runtime,  we see (A)  view created for first MyModule (and orphaned), (B) execute first compiled MyModule, using old view, ...

This is a problem, because we want backward hooks to run right after each compiled-backward, but autograd executes those hooks in an order mirroring their execution order during forward.  Since we are forever using the views created during steps (1, 3, ..  N), which all happen before the steps (A, B, ...),  this means that all the hooks will happen after all the compiled backwards.  An illustration of the problem - a torchviz graph showing the 2 possible orderings of autograd, and a profile showing the view-backwards ops happening after all the compiled backwards, and before all the backward hooks.

image
image

A solution is to make dynamo not specialize on these nn modules.  It is worth pointing out that this nn.module specialization is de-facto failing, as we are modifying .parameters and this bypasses dynamo's __setattr__ monkeypatch, which should have automatically kicked us out to Unspecialized and forced a recompile.

After unspecializing, the new views (created during steps A,  C, ...) are actually _used_ at runtime by the module, making their creation order interleaved, making autograd execute their backwards interleaved.

The new torchviz graph (this time with names added for the view tensors):
image

And a new profile showing the interleaving of compiled backwards and hooks, allowing overlapping of reduce-scatter.
image

@jansel @davidberard98 @aazzolini @mrshenli @awgu @ezyang @soumith @voznesenskym @anijain2305 

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by wconstab** [comment]
@jansel do you see any issues with landing this, if the perf for FSDP turns out ok? haven't benchmarked yet.  Will be taking a look at the impact of extra guards.  No idea if there are actually backend-optimizations on parameters i'm opting out of this way (for training).

@awgu what about the FSDP changes?  Any suggestion to make this less hacky? Any other places i need to add tagging to cover all cases?

We're not ready to land this yet anyway, going to do some deeper dives on perf/memory for a couple of models and also sweep a wider set of models to see if there are more gaps to fill on UnspecializedNNModule.  I recall previously i tried to switch all dynamo handling of nnModule to Unspec, and it failed a lot of tests, so i expect there would be more gaps in Unspec support.  @jansel is it right to consider filling these gaps as "fixing bugs" as Unspec should be able to handle fall back from any specialized NNMod?  Maybe we'll find it's better to refactor Unspec and NNModuleVariable to have a common base to avoid too much code dupe.

**Comment by aazzolini** [comment]
A few questions:
- why are we creating a new view at every iteration? Is it because the storage pointer to the flat buffer changes at every iteration?

- I see there's also a cuda graph disable change , is that a correctness thing or just performance?

File: torch/_dynamo/variables/nn_module.py [hunk 1]
```diff
@@ -567,7 +567,12 @@ def call_method(
                 return variables.ListIteratorVariable(
                     items, mutable_local=MutableLocal(), **options
                 )
-
+            elif isinstance(method, staticmethod):
+                return tx.inline_user_function_return(
+                    variables.UserFunctionVariable(method.__func__, **options),
```
Comment (aazzolini) [inline torch/_dynamo/variables/nn_module.py:572]: is this also part of the changes needed to "unspecialize" the module or is it something else?

File: benchmarks/dynamo/distributed.py [hunk 2]
```diff
@@ -81,6 +83,9 @@ def move_tensor(maybe_tensor):
             dynamo.config.log_level = logging.DEBUG
         if args.dynamo_optimize_ddp:
             dynamo.config.optimize_ddp = True
+        if args.dynamo == "inductor" and args.fsdp:
+            torch._inductor.config.triton.cudagraphs = False
+            log.warn("disabling inductor cudagraphs for compatibility with FSDP")
```
Comment (aazzolini) [inline benchmarks/dynamo/distributed.py:88]: is this related or something different?

**Review by aazzolini — COMMENTED** [review]
some questions inline. 

and another question, why do we have "specialized" nn modules in the first place?

File: torch/_dynamo/variables/nn_module.py (hunk 1 above)
Comment (wconstab) [reply torch/_dynamo/variables/nn_module.py:572]: this is more of a 'missing feature of UnspecializedNNModule that is now being exercised since we are directing more traffic to it'.  I expect there are more elif cases missing here that we'll find by testing more models.  But i also claim that UnspecializedNNModule is implicitly supposed to work for all cases, since any time we mutate an NNModule we will fall back to Unspecialized.

**Comment by awgu** [comment]
> * why are we creating a new view at every iteration? Is it because the storage pointer to the flat buffer changes at every iteration?

This is a good question! Yes, one reason is that the `FlatParameter` storage is allocated anew each iteration.

File: benchmarks/dynamo/distributed.py (hunk 2 above)
Comment (wconstab) [reply benchmarks/dynamo/distributed.py:88]: well, it's somewhat coupled.  There were weird bugs when I ran with cudagraphs, and i haven't debugged yet.  I think we agreed to make cudagraphs disabled by default in inductor, which would make this unnecessary.

**Comment by wconstab** [comment]
> * why are we creating a new view at every iteration? Is it because the storage pointer to the flat buffer changes at every iteration?

I believe the explanation above for the autograd behavior implies that we _must_ recreate the view at each iteration in order for the backward comm operations to run inbetween compute operations.  If autograd will follow the reverse order of creation when scheduling during backwards, the order we'd see is 
   1) all of backward compute ops (since they are created on iteration N)
   2) all of the accumulategrad for views (since views were created on iteration 0 and thus older than backward compute)
   3) backward hooks scheduling reduce scatters, not  overlapped with comm (since they run after accumulategrad)

This would be a pre-existing requirement,  not related to dynamo

> * I see there's also a cuda graph disable change , is that a correctness thing or just performance?

correctness/crash; but not root-caused

> and another question, why do we have "specialized" nn modules in the first place?

See https://github.com/pytorch/torchdynamo/issues/1895 for more background; it's a point of debate, but TLDR is it makes dynamo faster albeit sometimes sketchy.

File: torch/_dynamo/variables/builder.py [hunk 3]
```diff
@@ -297,7 +297,7 @@ def index_source(key):
                 return self.tx.output.side_effects.track_object_existing(
                     self.source, value, result
                 )
-            elif issubclass(
+            elif getattr(value, "_is_fsdp_wrapped_module", False) or issubclass(
                 value.__class__, torch.nn.parallel.distributed.DistributedDataParallel
             ):
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/_dynamo/variables/builder.py]: i don't remember. this  may have even been added by someone before i started working on DDP.

but, note, we are _unspecializing_ for DDP here.  not specializing.  so this is technically a 'safer' path to take.

File: torch/_dynamo/variables/builder.py [hunk 4]
```diff
@@ -298,7 +298,7 @@ def index_source(key):
                 return self.tx.output.side_effects.track_object_existing(
                     self.source, value, result
                 )
-            elif issubclass(
+            elif getattr(value, "_is_fsdp_managed_module", False) or issubclass(
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/_dynamo/variables/builder.py:301]: I stuck in some new comments :) good idea.

**Comment by jansel** [comment]
> @jansel do you see any issues with landing this, if the perf for FSDP turns out ok? haven't benchmarked yet. Will be taking a look at the impact of extra guards. No idea if there are actually backend-optimizations on parameters i'm opting out of this way (for training).

I am fine landing this to unblock FSDP.

> We're not ready to land this yet anyway, going to do some deeper dives on perf/memory for a couple of models and also sweep a wider set of models to see if there are more gaps to fill on UnspecializedNNModule. I recall previously i tried to switch all dynamo handling of nnModule to Unspec, and it failed a lot of tests, so i expect there would be more gaps in Unspec support. @jansel is it right to consider filling these gaps as "fixing bugs" as Unspec should be able to handle fall back from any specialized NNMod? Maybe we'll find it's better to refactor Unspec and NNModuleVariable to have a common base to avoid too much code dupe.

Yes, I would consider these bugfixes/enhancements.  I think there will still be a place for `NNModuleVariable` (weight prepacking optimizations etc), but ideally both codepaths would be robust.

**Comment by davidberard98** [comment]
fsdp_nov28_latency
fsdp_nov28_memory
Memory usage and latency ratio measurements, from the changes in this PR.

**Review by davidberard98 — APPROVED** [review]
LGTM!

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -330,8 +330,34 @@ def __init__(
     ):
         torch._C._log_api_usage_once("torch.distributed.fsdp")
         super().__init__()
-
         _init_ignored_module_states(self, module, ignored_modules)
+
+        # Add module annotations for Dynamo support
+        for submodule in module.modules():
+            if submodule not in self._ignored_modules:
+                """[Dynamo treats FSDP wrapped modules as UnspecializedNNModule]
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:338]: ok good point. i'll fix in a subsequent PR

**Comment by wconstab** [comment]
@pytochbot merge
